apps/playwright-browser-tunnel/src/PlaywrightBrowserTunnel.ts

```diff
@@ -0,0 +1,524 @@
+  private _ws?: WebSocket;
+  private _mode: ITunnelMode;
+  private readonly _wsEndpoint?: string;
+  private readonly _listenPort?: number;
+  private readonly _tmpPath: string;
+
+  public constructor(options: IPlaywrightTunnelOptions) {
+    const { mode, terminal, onStatusChange, tmpPath } = options;
+
+    if (mode === 'poll-connection') {
+      if (!options.wsEndpoint) {
+        throw new Error('wsEndpoint is required for poll-connection mode');
+      }
+      this._wsEndpoint = options.wsEndpoint;
+    } else if (mode === 'wait-for-incoming-connection') {
+      if (options.listenPort === undefined) {
+        throw new Error('listenPort is required for wait-for-incoming-connection mode');
+      }
+      this._listenPort = options.listenPort;
+    } else {
+      throw new Error(`Invalid mode: ${mode}`);
+    }
+
+    this._mode = mode;
+    this._terminal = terminal;
+    this._onStatusChange = onStatusChange;
+    this._tmpPath = tmpPath;
+  }
+
+  public get status(): TunnelStatus {
+    return this._status;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  private set status(newStatus: TunnelStatus) {
+    this._status = newStatus;
+    this._onStatusChange(newStatus);
+  }
+
+  public async waitForCloseAsync(): Promise<void> {
+    const terminal: ITerminal = this._terminal;
+    await new Promise<void>((resolve) => {
+      void this._initWsPromise?.then((ws) => {
+        ws.on('close', () => {
+          terminal.writeLine('WebSocket connection closed. resolving init promise.');
+          this._initWsPromise = undefined;
+          resolve();
+        });
+      });
+    });
+  }
+
+  public async startAsync(options: { keepRunning?: boolean } = {}): Promise<void> {
+    this._keepRunning = options.keepRunning ?? true;
+    const terminal: ITerminal = this._terminal;
+    terminal.writeLine(`keepRunning: ${this._keepRunning}`);
+    while (this._keepRunning) {
+      if (!this._initWsPromise) {
+        this._initWsPromise = this._initPlaywrightBrowserTunnelAsync();
+      } else {
+        terminal.writeLine(`Tunnel is already running with status: ${this.status}`);
+      }
+      await this.waitForCloseAsync();
+    }
+  }
+
+  public async stopAsync(): Promise<void> {
+    this._keepRunning = false;
+    void this._initWsPromise?.finally(() => {
+      this._ws?.close();
+    });
+  }
+
+  public async [Symbol.asyncDispose](): Promise<void> {
+    this._terminal.writeLine('Disposing WebSocket connection.');
+    await this.stopAsync();
+  }
+
+  public async cleanTempFilesAsync(): Promise<void> {
+    const tmpPath: string = this._tmpPath;
+    this._terminal.writeLine(`Cleaning up temporary files in ${tmpPath}`);
+    try {
+      await FileSystem.ensureEmptyFolderAsync(tmpPath);
+      this._terminal.writeLine(`Temporary files cleaned up.`);
+    } catch (error) {
+      this._terminal.writeLine(
+        `Failed to clean up temporary files: ${error instanceof Error ? error.message : error}`
```

Comment:
Please factor error stringification into a helper function and make it either always log stacks or be configurable to log stacks.